src/java/org/apache/cassandra/service/accord/AccordService.java

```diff
@@ -540,133 +508,234 @@ public IVerbHandler<? extends Reply> responseHandler()
+                        // The caller can retry if the table indeed exists and is managed by Accord.
+                        Set<TableId> txnDroppedTables = txnDroppedTables(keysOrRanges);
+                        Tracing.trace("Accord returned topology mismatch: " + coordinationFailed.getMessage());
+                        logger.debug("Accord returned topology mismatch", coordinationFailed);
+                        bookkeeping.markTopologyMismatch();
+                        // Throw IRE in case the caller fails to check if the table still exists
+                        if (!txnDroppedTables.isEmpty())
+                        {
+                            Tracing.trace("Accord txn uses dropped tables {}", txnDroppedTables);
+                            logger.debug("Accord txn uses dropped tables {}", txnDroppedTables);
+                            throw new InvalidRequestException("Accord transaction uses dropped tables");
+                        }
+                        trySuccess((V)RetryWithNewProtocolResult.instance);
+                        return;
+                    }
+                    else
+                    {
+                        // this case happens when a non-timeout exception is seen, and we are unable to move forward
+                        if (coordinationFailed instanceof Exhausted)
+                            report = bookkeeping.newExhausted(txnId, keysOrRanges);
+                        else
+                            report = bookkeeping.newFailed(txnId, keysOrRanges);
+                    }
+                    if (txnId != null && txnId.isSyncPoint())
+                        AccordAgent.onFailedBarrier(txnId, fail);
+                }
+                else
+                {
+                    report = bookkeeping.newFailed(txnId, keysOrRanges);
+                }
+                report.addSuppressed(fail);
+                tryFailure(report);
             }
-            if (cause instanceof Exhausted)
+            else
             {
-                TxnId txnId = ((Exhausted) cause).txnId();
-                ((AccordAgent) node.agent()).onFailedBarrier(txnId, keysOrRanges, cause);
-                // this case happens when a non-timeout exception is seen, and we are unable to move forward
-                metrics.failures.mark();
-                throw newBarrierExhausted(((CoordinationFailed)cause).txnId(), barrierType, isForWrite, keysOrRanges);
+                if (success == RetryWithNewProtocolResult.instance)
+                {
+                    bookkeeping.markRetryDifferentSystem();
+                    Tracing.trace("Got retry different system error from Accord, will retry");
+                }
+                trySuccess(success);
             }
-            // unknown error
-            metrics.failures.mark();
-            throw new RuntimeException(cause);
+            bookkeeping.markElapsedNanos(nanoTime() - startedAtNanos);
```

Comment:
Do we also want to report latency in the failed case of topology mismatch above?

Reply:
If you mean the retry on another system event then no, I don't think we should - the total latency for the request will be reported against the final system that's chosen, and we'll double count otherwise (it looks like we already double count for CAS loops, but that's a separate problem)